cache/estargz.go

```diff
@@ -11,76 +11,117 @@ import (
+				// same tar as the original.
+				//
+				// Note that we don't support eStragz compression for tar that contains a file named
+				// `stargz.index.json` because we cannot create eStargz in loseless way for such blob
+				// (we must overwrite stargz.index.json file).
+				if err := w.AppendTarLossLess(pr); err != nil {
 					pr.CloseWithError(err)
-					return
+					return err
 				}
 				tocDgst, err := w.Close()
 				if err != nil {
 					pr.CloseWithError(err)
-					return
+					return err
 				}
-				if err := cw.Close(); err != nil {
+				if err := blobInfoW.Close(); err != nil {
 					pr.CloseWithError(err)
-					return
+					return err
 				}
+				bInfo := <-bInfoCh
 				mu.Lock()
-				bInfo = <-bInfoCh
-				annotations[estargz.TOCJSONDigestAnnotation] = tocDgst.String()
-				annotations[estargz.StoreUncompressedSizeAnnotation] = fmt.Sprintf("%d", bInfo.uncompressedSize)
+				cInfo = &compressionInfo{bInfo, tocDgst}
 				mu.Unlock()
+				pr.Close()
+				return nil
 			}()
 			return &writeCloser{pw, func() error {
 				<-done // wait until the write completes
 				return nil
 			}}, nil
 		}, func(ctx context.Context, cs content.Store) (map[string]string, error) {
-			a := make(map[string]string)
 			mu.Lock()
-			bInfo := bInfo
-			for k, v := range annotations {
-				a[k] = v
-			}
+			cInfo, writeErr := cInfo, writeErr
 			mu.Unlock()
-			info, err := cs.Info(ctx, bInfo.compressedDigest)
+			if cInfo == nil {
+				if writeErr != nil {
+					return nil, errors.Wrapf(writeErr, "cannot finalize due to write error")
+				}
+				return nil, errors.Errorf("cannot finalize (reason unknown)")
+			}
+
+			// Fill necessary labels
+			info, err := cs.Info(ctx, cInfo.compressedDigest)
 			if err != nil {
 				return nil, errors.Wrap(err, "failed to get info from content store")
 			}
 			if info.Labels == nil {
 				info.Labels = make(map[string]string)
 			}
-			info.Labels[containerdUncompressed] = bInfo.uncompressedDigest.String()
+			info.Labels[containerdUncompressed] = cInfo.uncompressedDigest.String()
 			if _, err := cs.Update(ctx, info, "labels."+containerdUncompressed); err != nil {
 				return nil, err
 			}
-			a[containerdUncompressed] = bInfo.uncompressedDigest.String()
+
+			// Fill annotations
+			a := make(map[string]string)
+			a[estargz.TOCJSONDigestAnnotation] = cInfo.tocDigest.String()
+			a[estargz.StoreUncompressedSizeAnnotation] = fmt.Sprintf("%d", cInfo.uncompressedSize)
+			a[containerdUncompressed] = cInfo.uncompressedDigest.String()
 			return a, nil
 		}
 }
 
+// isEStargz returns true when the specified digest of content exists in
+// the content store as eStargz.
+func isEStargz(ctx context.Context, cs content.Store, dgst digest.Digest) bool {
+	r, err := cs.ReaderAt(ctx, ocispecs.Descriptor{Digest: dgst})
+	if err != nil {
+		return false
+	}
+	defer r.Close()
+	_, _, err = estargz.OpenFooter(io.NewSectionReader(r, 0, r.Size()))
```

Comment:
Could we leave a label on the blob after this has been determined for the first time and then avoid opening blob after that